Sync/BatchingClient.swift

```diff
@@ -0,0 +1,221 @@
+
+public class Sync15BatchClient<T: CleartextPayloadJSON> {
+    private let config: InfoConfiguration
+    private let uploader: BatchUploadFunction
+    private let serializeRecord: (Record<T>) -> String?
+    private var ifUnmodifiedSince: Timestamp?
+
+    private var batchMeta: BatchMeta
+    private var batchToken: BatchToken?
+    private var currentPayload: Payload
+    private var onCollectionUploaded: (POSTResult -> Void)?
+
+    // TODO: use I-U-S.
+    // Each time we do the storage operation, we might receive a backoff notification.
+    // For a success response, this will be on the subsequent request, which means we don't
+    // have to worry about handling successes and failures mixed with backoffs here.
+
+    init(config: InfoConfiguration, ifUnmodifiedSince: Timestamp? = nil, serializeRecord: (Record<T>) -> String?, uploader: BatchUploadFunction, onCollectionUploaded: (POSTResult -> Void)?) {
+        self.config = config
+        self.ifUnmodifiedSince = ifUnmodifiedSince
+        self.uploader = uploader
+        self.serializeRecord = serializeRecord
+
+        self.batchMeta = BatchMeta(config: config)
+        self.currentPayload = Payload(config: config)
+        self.onCollectionUploaded = onCollectionUploaded
+    }
+
+    public func addRecord(record: Record<T>) -> Success {
+        // Sanity checks and assumptions
+        precondition(config.maxPostBytes < config.maxTotalBytes)
+        precondition(config.maxPostRecords < config.maxTotalRecords)
+
+        guard let line = self.serializeRecord(record) else {
+            return deferMaybe(SerializeRecordFailure(record: record))
+        }
+
+        let lineSize = line.utf8.count
+        guard lineSize < Sync15StorageClient.maxPayloadSizeBytes else {
+            return deferMaybe(RecordTooLargeError(size: lineSize, guid: record.id))
+        }
+
+        let uploadRecord: UploadRecord = (record.id, line, lineSize)
+
+        // If we can add a record to the payload, all is good - we don't need to push to the server yet.
+        // Otherwise, try to push the payload (since it's full) to the server in a batch
+        if self.currentPayload.add(uploadRecord) {
+            return succeed()
+        }
+
+        // If we have a batch token, go ahead and push up the payload using the token.
+        // Otherwise, attempt to start a new batch. If we fail starting - no problem! The records would
+        // have already been sent in a single POST call.
+        guard let token = batchToken else {
+            return self.start(self.currentPayload)
+                >>== effect({ self.batchToken = $0 })
+
+                // Based on the precondition, we must be able to add at least one payload into the batch 
+                // without it failing
+                >>> effect({ self.batchMeta.add(self.currentPayload) })
+                >>> effect({ self.currentPayload = Payload(config: self.config) })
```

Comment:
Also Grisha pointed out on IRC that there is a bug here that the record never gets added to the new payload. I've added a fix for that.